ludwig/utils/trainer_utils.py

```diff
@@ -191,8 +198,14 @@ def __init__(
         self.best_eval_validation_metrics = best_eval_validation_metrics
         self.best_eval_test_metrics = best_eval_test_metrics
 
+        # Token usage.
+        self.incremental_token_usage = incremental_token_usage
+        self.cumulative_token_usage = cumulative_token_usage
+        self.total_tokens_used = total_tokens_used
+
     def save(self, filepath):
-        save_json(filepath, self.__dict__)
+        # sort_keys=False to ensure that token usage dictionaries (keyed by integers) are encodable.
+        save_json(filepath, self.__dict__, sort_keys=False)
```

Comment:
Oh super interesting - how has this worked historically with current epochs/steps? Would it perhaps makes sense to change the dictionary mapping to `Dict[str, int]` instead of `Dict[int,int]` so that we don't have to set `sort_keys=False`? Might also help with consistency with metrics stored per epoch which I believe are keyed by dataset (train, test, val) + metric + epoch, all of which are strings?

Reply:
It turns out that we do want to index by strings because deserializing from JSON when restarting from an existing checkpoint, any dictionary's keys get automatically serialized as strings anyway. I've added a comment about this in the ProgressTracker docstring as this behavior is not at all obvious.